### PR TITLE
Removed duplicate data and added sol transfers

### PR DESCRIPTION
**Pull request by matzayonc**

### CODE REVIEW

File: programs/protocol/src/actions/forwarder/make_offer.rs [hunk 1]
```diff
@@ -33,30 +33,31 @@ pub struct MakeOffer<'info> {
...
     let shipment = &mut ctx.accounts.shipment.load_mut()?;
 
-    require_eq!(shipment.buyer, shipment.owner, Error::ShipmentSold);
+    require_eq!(shipment.carrier, Pubkey::default(), Error::ShipmentSold);
 
     let now = Clock::get()?.unix_timestamp;
 
     **offer = ShipmentOffer {
-        owner: carrier.creator,
+        offeror: forwarder.creator,
```
Comment (tommysr) [inline programs/protocol/src/actions/forwarder/make_offer.rs:44]: seems like error

File: programs/protocol/src/actions/forwarder/buy_shipment.rs [hunk 2]
```diff
@@ -25,24 +25,37 @@ pub struct BuyShipment<'info> {
...
 
-    **forwarded = BoughtShipment {
-        creator: shipper.creator,
-        buyer: ctx.accounts.signer.key(),
-        owner: ctx.accounts.signer.key(),
+    // Create forwarded shipment
+    **forwarded = ForwardedShipment {
+        forwarder: forwarder.creator,
+        shipment: ctx.accounts.shipment.key(),
+        resell_price: u64::MAX,
```
Comment (tommysr) [inline programs/protocol/src/actions/forwarder/buy_shipment.rs:56]: why introduce it if its placeholder

File: programs/protocol/src/actions/carrier/accept_offer.rs [hunk 3]
```diff
@@ -14,13 +12,13 @@ pub struct AcceptOffer<'info> {
     pub task: AccountLoader<'info, AcceptedOffer>,
     #[account(mut,
         seeds = [b"offer", carrier.load().unwrap().creator.as_ref(), &offer.load().unwrap().no.to_le_bytes()], bump,
-        // close = signer,
+        close = signer,
```
Comment (tommysr) [inline programs/protocol/src/actions/carrier/accept_offer.rs:15]: why not keep it for stats?

Comment (matzayonc) [reply programs/protocol/src/actions/carrier/accept_offer.rs:15]: I don't think there is any data deleted here

**Review by tommysr — APPROVED** [review]
A lot of structural changes. Those were described as helper for front, but they're now nightmare for front.

File: programs/protocol/src/actions/carrier/accept_offer.rs (hunk 3 above)
Comment (tommysr) [reply programs/protocol/src/actions/carrier/accept_offer.rs:15]: but test say the other way

File: programs/protocol/src/actions/forwarder/buy_shipment.rs (hunk 2 above)
Comment (matzayonc) [reply programs/protocol/src/actions/forwarder/buy_shipment.rs:56]: seems like might be useful on the frontend

Comment (tommysr) [reply programs/protocol/src/actions/forwarder/buy_shipment.rs:56]: how

**Review by tejks — APPROVED** [review]

**Comment by tommysr** [comment]
@matzayonc i would be glad to have it deployed on master

File: programs/protocol/src/actions/carrier/accept_offer.rs (hunk 3 above)
Comment (tommysr) [reply programs/protocol/src/actions/carrier/accept_offer.rs:15]: in test you're asserting null after accepting the offer. can u explain?

File: programs/protocol/src/actions/forwarder/buy_shipment.rs (hunk 2 above)
Comment (matzayonc) [reply programs/protocol/src/actions/forwarder/buy_shipment.rs:56]: to not change it everywhere when it is implemented

**Comment by tommysr** [comment]
@matzayonc bump

**Comment by matzayonc** [comment]
> @matzayonc bump
> 

> is it not?

File: programs/protocol/src/data/channel.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+use anchor_lang::prelude::{
+    borsh::{BorshDeserialize, BorshSerialize},
+    *,
+};
+use uint::construct_uint;
+
+use crate::Name;
+
+construct_uint! {
+    pub struct Key(4); // 4 * 64bit = 256 bit
```
Comment (tommysr) [inline programs/protocol/src/data/channel.rs:10]: do you even use it?

**Review by tommysr — APPROVED** [review]
i would search for ready applications, which allow to use some shared secret key, but looks good for me.

**Comment by tommysr** [comment]
please resolve conflicts, merge and deploy

**Pull request merged**